File: App2/core/data/src/main/java/com/app2/core/data/remote/dto/POSDTOs.kt
package com.app2.core.data.remote.dto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class OpenSessionResponse(
    val success: Boolean = true,
    val session: POSSessionDTO? = null,
    @SerialName("session_number") val sessionNumber: String? = null
)

@Serializable
data class CreateCashMovementResponse(
    val success: Boolean = true,
    val movement: CashMovementDTO? = null
)

@Serializable
data class BestSellerDTO(
    val id: Int,
    val name: String,
    val price: Double = 0.0,
    @SerialName("price_base") val priceBase: Double? = null,
    val sku: String = "",
    val quantity: Int? = null,
    @SerialName("total_sold") val totalSold: Int = 0
)

@Serializable
data class POSSessionDTO(
    val id: Int,
    @SerialName("session_number") val sessionNumber: String? = null,
    @SerialName("warehouse_id") val warehouseId: Int? = null,
    @SerialName("opening_cash") val openingCash: Double? = null,
    @SerialName("closing_cash") val closingCash: Double? = null,
    @SerialName("expected_cash") val expectedCash: Double? = null,
    val status: String? = null,
    @SerialName("opened_at") val openedAt: String? = null,
    @SerialName("closed_at") val closedAt: String? = null
)

@Serializable
data class CloseSessionResponse(
    val success: Boolean = true,
    @SerialName("expected_cash") val expectedCash: Double? = null,
    val deposited: Boolean? = null
)

@Serializable
data class CashMovementDTO(
    val id: Int,
    @SerialName("session_id") val sessionId: Int? = null,
    val type: String? = null,
    val amount: Double? = null,
    val reason: String? = null,
    val note: String? = null,
    @SerialName("created_at") val createdAt: String? = null
)

@Serializable
data class POSTransactionDTO(
    val id: Int,
    @SerialName("ticket_number") val ticketNumber: String? = null,
    @SerialName("transaction_number") val transactionNumber: String? = null,
    @SerialName("session_id") val sessionId: Int? = null,
    @SerialName("customer_id") val customerId: Int? = null,
    @SerialName("customer_name") val customerName: String? = null,
    @SerialName("payment_method") val paymentMethod: String? = null,
    val subtotal: Double? = null,
    @SerialName("discount_total") val discountTotal: Double? = null,
    @SerialName("tax_amount") val taxAmount: Double? = null,
    val total: Double? = null,
    @SerialName("tendered_amount") val tenderedAmount: Double? = null,
    @SerialName("change_given") val changeGiven: Double? = null,
    val status: String? = null,
    @SerialName("created_at") val createdAt: String? = null
)

@Serializable
data class POSCustomerDTO(
    val id: Int,
    val name: String,
    @SerialName("client_code") val clientCode: String? = null,
    @SerialName("discount_rate") val discountRate: Double? = null
)

@Serializable
data class POSTransactionRequest(
    @SerialName("session_id") val sessionId: Int,
    @SerialName("customer_id") val customerId: Int? = null,
    val items: List<TransactionItem>,
    @SerialName("payment_method") val paymentMethod: String,
    @SerialName("tendered_amount") val tenderedAmount: Double,
    val notes: String = "",
    @SerialName("is_credit") val isCredit: Boolean = false
) {
    @Serializable
    data class TransactionItem(
        @SerialName("product_id") val productId: Int,
        val quantity: Int,
        @SerialName("unit_price") val unitPrice: Double,
        @SerialName("discount_percent") val discountPercent: Double = 0.0,
        @SerialName("product_name") val productName: String = "",
        @SerialName("product_sku") val productSku: String = ""
    )
}

@Serializable
data class POSTransactionResponse(
    val success: Boolean = true,
    @SerialName("document_number") val documentNumber: String? = null,
    @SerialName("document_id") val documentId: Int? = null,
    @SerialName("document_type") val documentType: String? = null,
    @SerialName("document_status") val documentStatus: String? = null,
    val total: Double? = null,
    @SerialName("change_amount") val changeAmount: Double? = null,
    @SerialName("customer_name") val customerName: String? = null
)

@Serializable
enum class DiscountTier {
    @SerialName("normal") Normal,
    @SerialName("loyal") Loyal,
    @SerialName("student") Student,
    @SerialName("school") School;

    val label: String
        get() = when (this) {
            Normal -> "Normal"
            Loyal -> "Fidèle"
            Student -> "Étudiant"
            School -> "École"
        }
}

@Serializable
enum class PaymentMethod {
    @SerialName("cash") Cash,
    @SerialName("card") Card,
    @SerialName("mixed") Mixed;

    val label: String
        get() = when (this) {
            Cash -> "Espèces"
            Card -> "Carte bancaire"
            Mixed -> "Mixte"
        }
}

@Serializable
data class CartItem(
    @SerialName("product_id") val productId: Int,
    @SerialName("product_name") val productName: String,
    @SerialName("product_sku") val productSku: String,
    val quantity: Int = 1,
    @SerialName("unit_price") val unitPrice: Double,
    @SerialName("base_unit_price") val baseUnitPrice: Double,
    @SerialName("price_loyal") val priceLoyal: Double? = null,
    @SerialName("price_school") val priceSchool: Double? = null,
    @SerialName("price_student") val priceStudent: Double? = null,
    @SerialName("discount_percent") val discountPercent: Double = 0.0
)

File: ios/StockPro/Data/DTOs/POSDTOs.swift
import Foundation

struct OpenSessionResponse: Decodable {
    let success: Bool
    let session: POSSessionDTO?
    let session_number: String?
}

struct CreateCashMovementResponse: Decodable {
    let success: Bool
    let movement: CashMovementDTO?
}

struct CartItem: Identifiable, Equatable {
    let id = UUID()
    let productId: Int
    let productName: String
    let productSku: String
    var quantity: Int
    var unitPrice: Double
    var baseUnitPrice: Double
    var priceLoyal: Double?
    var priceSchool: Double?
    var priceStudent: Double?
    var discountPercent: Double

    var lineTotal: Double {
        let base = Double(quantity) * unitPrice
        return base - (base * discountPercent / 100)
    }

    var taxAmount: Double { lineTotal * 0.20 }
    var formattedUnitPrice: String { String(format: "%.2f MAD", unitPrice) }
    var formattedLineTotal: String { String(format: "%.2f MAD", lineTotal) }
}

struct BestSellerDTO: Decodable, Identifiable {
    let id: Int
    let name: String
    let price: Double
    let price_base: Double
    let sku: String
    let quantity: Int?
    let total_sold: Int

    var stockStatus: StockStatus {
        guard let q = quantity, q > 0 else { return .outOfStock }
        return .inStock
    }
}

struct POSSessionDTO: Decodable, Identifiable {
    let id: Int
    let session_number: String?
    let warehouse_id: Int?
    let opening_cash: Double?
    let closing_cash: Double?
    let expected_cash: Double?
    let status: String?
    let opened_at: String?
    let closed_at: String?

    var isOpen: Bool { status == "open" }
}

struct CloseSessionResponse: Decodable {
    let success: Bool
    let expected_cash: Double?
    let deposited: Bool?
}

struct CashMovementDTO: Decodable, Identifiable {
    let id: Int
    let session_id: Int?
    let type: String?
    let amount: Double?
    let reason: String?
    let note: String?
    let created_at: String?
}

struct POSTransactionDTO: Decodable, Identifiable {
    let id: Int
    let ticket_number: String?
    let transaction_number: String?
    let session_id: Int?
    let customer_id: Int?
    let customer_name: String?
    let payment_method: String?
    let subtotal: Double?
    let discount_total: Double?
    let tax_amount: Double?
    let total: Double?
    let tendered_amount: Double?
    let change_given: Double?
    let status: String?
    let created_at: String?
}

struct POSCustomerDTO: Decodable, Identifiable {
    let id: Int
    let name: String
    let client_code: String?
    let discount_rate: Double?
}

struct POSTransactionRequest: Encodable {
    let session_id: Int
    let customer_id: Int?
    let items: [TransactionItem]
    let payment_method: String
    let tendered_amount: Double
    let notes: String
    let is_credit: Bool

    struct TransactionItem: Encodable {
        let product_id: Int
        let quantity: Int
        let unit_price: Double
        let discount_percent: Double
        let product_name: String
        let product_sku: String
    }
}

struct POSTransactionResponse: Decodable {
    let success: Bool
    let document_number: String?
    let document_id: Int?
    let document_type: String?
    let document_status: String?
    let total: Double?
    let change_amount: Double?
    let customer_name: String?
}

enum DiscountTier: String, CaseIterable {
    case normal = "normal"
    case loyal = "loyal"
    case student = "student"
    case school = "school"

    var label: String {
        switch self {
        case .normal: "Normal"
        case .loyal: "Fidèle"
        case .student: "Étudiant"
        case .school: "École"
        }
    }

    private var fallbackRate: Double {
        switch self {
        case .normal: return 0
        case .loyal: return 0.15
        case .student: return 0.15
        case .school: return 0.20
        }
    }

    func priceFor(item: CartItem) -> Double {
        switch self {
        case .normal: return item.baseUnitPrice
        case .loyal: return item.priceLoyal ?? item.baseUnitPrice * (1 - 0.15)
        case .student: return item.priceStudent ?? item.baseUnitPrice * (1 - 0.15)
        case .school: return item.priceSchool ?? item.baseUnitPrice * (1 - 0.20)
        }
    }
}

enum PaymentMethod: String, CaseIterable {
    case cash = "cash"
    case card = "card"
    case mixed = "mixed"

    var label: String {
        switch self {
        case .cash: "Espèces"
        case .card: "Carte bancaire"
        case .mixed: "Mixte"
        }
    }

    var icon: String {
        switch self {
        case .cash: "banknote"
        case .card: "creditcard"
        case .mixed: "rectangle.split.2x2"
        }
    }
}
